File: src/services/userPreferenceLearning.js
/**
 * User Preference Learning System
 * Implements a feedback loop to continuously learn and adapt to user preferences
 * Uses reinforcement learning principles to improve recommendations over time
 */

export class UserPreferenceLearning {
  constructor(mlEngine) {
    this.mlEngine = mlEngine;
    this.learningRate = 0.1;
    this.feedbackHistory = this.loadFeedbackHistory();
    this.sessionData = [];
  }

  /**
   * Record user feedback on a recommendation
   * @param {Object} recommendation - The recommendation that was shown
   * @param {string} feedback - 'like', 'dislike', 'worn', 'ignored'
   * @param {Object} weatherContext - Weather conditions when feedback was given
   */
  recordFeedback(recommendation, feedback, weatherContext) {
    const feedbackEntry = {
      timestamp: Date.now(),
      recommendation: {
        item: recommendation.item,
        category: recommendation.category,
        confidence: recommendation.confidence,
      },
      feedback,
      weatherContext: {
        temperature: weatherContext.temperature,
        condition: weatherContext.condition,
        timeOfDay: weatherContext.timeOfDay,
      },
      sessionId: this.getCurrentSessionId(),
    };

    this.feedbackHistory.push(feedbackEntry);
    this.sessionData.push(feedbackEntry);

    // Update ML weights based on feedback
    this.updateModelWeights(feedbackEntry);

    // Update user profile
    this.updateUserProfile(feedbackEntry);

    // Save to storage
    this.saveFeedbackHistory();
    this.mlEngine.saveModelWeights();
    this.mlEngine.saveUserProfile();

    return feedbackEntry;
  }

  /**
   * Update ML model weights using reinforcement learning
   */
  updateModelWeights(feedbackEntry) {
    const { recommendation, feedback, weatherContext } = feedbackEntry;

    // Calculate reward signal
    const reward = this.calculateReward(feedback);

    // Extract features
    const features = {
      temperature: this.normalizeTemperature(weatherContext.temperature),
      isRainy: weatherContext.condition?.toLowerCase().includes('rain') ? 1 : 0,
    };

    // Update weight for this specific recommendation pattern
    const featureKey = `${recommendation.category}_${Math.floor(features.temperature * 10)}_${features.isRainy}`;
    const currentWeight = this.mlEngine.modelWeights[featureKey] || 1;

    // Gradient descent update
    const newWeight = currentWeight + this.learningRate * reward;
    this.mlEngine.modelWeights[featureKey] = Math.max(0, Math.min(2, newWeight)); // Clamp between 0 and 2

    // Also update related patterns (generalization)
    this.updateRelatedWeights(featureKey, reward * 0.5);
  }

  /**
   * Calculate reward signal from user feedback
   */
  calculateReward(feedback) {
    switch (feedback) {
      case 'worn':
        return 1.0; // Strongest positive signal
      case 'like':
        return 0.5;
      case 'ignored':
        return -0.2; // Mild negative signal
      case 'dislike':
        return -0.5;
      case 'inappropriate':
        return -1.0; // Strongest negative signal
      default:
        return 0;
    }
  }

  /**
   * Update related weights for generalization
   */
  updateRelatedWeights(featureKey, adjustedReward) {
    const [category, tempBucket, isRainy] = featureKey.split('_');

    // Update adjacent temperature buckets
    const tempNum = parseInt(tempBucket);
    for (let offset = -1; offset <= 1; offset++) {
      if (offset === 0) continue;

      const relatedKey = `${category}_${tempNum + offset}_${isRainy}`;
      const currentWeight = this.mlEngine.modelWeights[relatedKey] || 1;
      const decay = Math.abs(offset) === 1 ? 0.5 : 0.25;

      this.mlEngine.modelWeights[relatedKey] = Math.max(
        0,
        Math.min(2, currentWeight + this.learningRate * adjustedReward * decay)
      );
    }
  }

  /**
   * Update user profile with preference learning
   */
  updateUserProfile(feedbackEntry) {
    const { recommendation, feedback } = feedbackEntry;
    const profile = this.mlEngine.userProfile;

    // Update category preferences
    if (!profile.categoryPreferences) profile.categoryPreferences = {};
    const categoryPref = profile.categoryPreferences[recommendation.category] || 1;

    const categoryReward = this.calculateReward(feedback) * 0.5;
    profile.categoryPreferences[recommendation.category] = Math.max(
      0,
      Math.min(2, categoryPref + this.learningRate * categoryReward)
    );

    // Update item preferences
    if (!profile.itemPreferences) profile.itemPreferences = {};
    const itemPref = profile.itemPreferences[recommendation.item] || 1;

    const itemReward = this.calculateReward(feedback);
    profile.itemPreferences[recommendation.item] = Math.max(
      0,
      Math.min(2, itemPref + this.learningRate * itemReward)
    );

    // Store feedback in profile
    if (!profile.feedbackHistory) profile.feedbackHistory = [];
    profile.feedbackHistory.push({
      timestamp: feedbackEntry.timestamp,
      item: recommendation.item,
      feedback,
    });

    // Keep only last 200 feedback entries
    if (profile.feedbackHistory.length > 200) {
      profile.feedbackHistory = profile.feedbackHistory.slice(-200);
    }
  }

  /**
   * Get personalized insights based on learning
   */
  getPersonalizedInsights() {
    if (this.feedbackHistory.length < 10) {
      return {
        dataPoints: this.feedbackHistory.length,
        message: 'Keep providing feedback to get personalized insights!',
        insights: [],
      };
    }

    const insights = [];
    const profile = this.mlEngine.userProfile;

    // Analyze category preferences
    const categoryPrefs = profile.categoryPreferences || {};
    const topCategories = Object.entries(categoryPrefs)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 3);

    if (topCategories.length > 0) {
      insights.push({
        type: 'preference',
        title: 'Your Favorite Categories',
        description: topCategories
          .map(([cat, score]) => `${cat} (${Math.round(score * 50)}% preference)`)
          .join(', '),
      });
    }

    // Analyze item preferences
    const itemPrefs = profile.itemPreferences || {};
    const lovedItems = Object.entries(itemPrefs)
      .filter(([, score]) => score > 1.3)
      .map(([item]) => item);

    if (lovedItems.length > 0) {
      insights.push({
        type: 'loved',
        title: 'Items You Love',
        description: lovedItems.slice(0, 5).join(', '),
      });
    }

    const dislikedItems = Object.entries(itemPrefs)
      .filter(([, score]) => score < 0.7)
      .map(([item]) => item);

    if (dislikedItems.length > 0) {
      insights.push({
        type: 'disliked',
        title: 'Items You Avoid',
        description: dislikedItems.slice(0, 5).join(', '),
      });
    }

    // Analyze weather patterns
    const weatherPatterns = this.analyzeWeatherPatterns();
    if (weatherPatterns) {
      insights.push(weatherPatterns);
    }

    return {
      dataPoints: this.feedbackHistory.length,
      message: `We've learned from ${this.feedbackHistory.length} of your feedback entries!`,
      insights,
      learningProgress: this.calculateLearningProgress(),
    };
  }

  /**
   * Analyze user's weather-based patterns
   */
  analyzeWeatherPatterns() {
    const recentFeedback = this.feedbackHistory.slice(-50);

    const coldWeatherFeedback = recentFeedback.filter(
      f => f.weatherContext.temperature < 10
    );

    const warmWeatherFeedback = recentFeedback.filter(
      f => f.weatherContext.temperature > 20
    );

    if (coldWeatherFeedback.length > 5) {
      const coldLikes = coldWeatherFeedback.filter(
        f => f.feedback === 'like' || f.feedback === 'worn'
      ).length;

      const coldPreference = coldLikes / coldWeatherFeedback.length;

      if (coldPreference > 0.7) {
        return {
          type: 'weather_pattern',
          title: 'Cold Weather Preference',
          description: 'You seem to appreciate our cold weather recommendations!',
        };
      }
    }

    if (warmWeatherFeedback.length > 5) {
      const warmLikes = warmWeatherFeedback.filter(
        f => f.feedback === 'like' || f.feedback === 'worn'
      ).length;

      const warmPreference = warmLikes / warmWeatherFeedback.length;

      if (warmPreference > 0.7) {
        return {
          type: 'weather_pattern',
          title: 'Warm Weather Preference',
          description: 'You love our warm weather recommendations!',
        };
      }
    }

    return null;
  }

  /**
   * Calculate overall learning progress
   */
  calculateLearningProgress() {
    const totalFeedback = this.feedbackHistory.length;

    if (totalFeedback < 10) return { level: 'beginner', percentage: 10 };
    if (totalFeedback < 30) return { level: 'learning', percentage: 30 };
    if (totalFeedback < 50) return { level: 'intermediate', percentage: 50 };
    if (totalFeedback < 100) return { level: 'advanced', percentage: 70 };
    return { level: 'expert', percentage: 90 };
  }

  /**
   * Get session-based analytics
   */
  getSessionAnalytics() {
    const totalRecommendations = this.sessionData.length;

    if (totalRecommendations === 0) {
      return null;
    }

    const liked = this.sessionData.filter(
      f => f.feedback === 'like' || f.feedback === 'worn'
    ).length;

    const disliked = this.sessionData.filter(
      f => f.feedback === 'dislike' || f.feedback === 'inappropriate'
    ).length;

    return {
      totalRecommendations,
      liked,
      disliked,
      ignored: totalRecommendations - liked - disliked,
      satisfactionRate: liked / totalRecommendations,
    };
  }

  /**
   * Reset session data (call when app restarts)
   */
  resetSession() {
    this.sessionData = [];
  }

  /**
   * Export user data for analysis or backup
   */
  exportUserData() {
    return {
      feedbackHistory: this.feedbackHistory,
      userProfile: this.mlEngine.userProfile,
      modelWeights: this.mlEngine.modelWeights,
      exportDate: new Date().toISOString(),
      dataPoints: this.feedbackHistory.length,
    };
  }

  /**
   * Import user data (for restore or migration)
   */
  importUserData(data) {
    if (data.feedbackHistory) {
      this.feedbackHistory = data.feedbackHistory;
      this.saveFeedbackHistory();
    }

    if (data.userProfile) {
      this.mlEngine.userProfile = data.userProfile;
      this.mlEngine.saveUserProfile();
    }

    if (data.modelWeights) {
      this.mlEngine.modelWeights = data.modelWeights;
      this.mlEngine.saveModelWeights();
    }

    return {
      success: true,
      dataPoints: this.feedbackHistory.length,
    };
  }

  /**
   * Clear all learning data (reset to defaults)
   */
  clearAllData() {
    this.feedbackHistory = [];
    this.sessionData = [];
    this.mlEngine.userProfile = {
      categoryPreferences: {},
      itemPreferences: {},
      feedbackHistory: [],
    };
    this.mlEngine.modelWeights = {};

    this.saveFeedbackHistory();
    this.mlEngine.saveUserProfile();
    this.mlEngine.saveModelWeights();

    return { success: true, message: 'All learning data cleared' };
  }

  // Helper functions
  getCurrentSessionId() {
    if (!this._sessionId) {
      this._sessionId = `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }
    return this._sessionId;
  }

  normalizeTemperature(temp) {
    return (temp + 10) / 50;
  }

  loadFeedbackHistory() {
    try {
      const history = localStorage.getItem('user_feedback_history');
      return history ? JSON.parse(history) : [];
    } catch (error) {
      console.error('Error loading feedback history:', error);
      return [];
    }
  }

  saveFeedbackHistory() {
    try {
      localStorage.setItem('user_feedback_history', JSON.stringify(this.feedbackHistory));
    } catch (error) {
      console.error('Error saving feedback history:', error);
    }
  }
}

File: backend/app/services/user_preference_learning.py
"""
User Preference Learning System
Implements a feedback loop to continuously learn and adapt to user preferences
Uses reinforcement learning principles to improve recommendations over time
"""
import uuid
from datetime import datetime
from typing import Dict, List, Any, Optional


class UserPreferenceLearning:
    def __init__(self, ml_engine, feedback_history: List[Dict] = None):
        self.ml_engine = ml_engine
        self.learning_rate = 0.1
        self.feedback_history = feedback_history or []
        self.session_data = []
        self._session_id = None

    def record_feedback(
        self,
        recommendation: Dict[str, Any],
        feedback: str,
        weather_context: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Record user feedback on a recommendation
        feedback: 'like', 'dislike', 'worn', 'ignored', 'inappropriate'
        """
        feedback_entry = {
            'timestamp': datetime.now().timestamp() * 1000,
            'recommendation': {
                'item': recommendation['item'],
                'category': recommendation['category'],
                'confidence': recommendation.get('confidence', 0)
            },
            'feedback': feedback,
            'weatherContext': {
                'temperature': weather_context['temperature'],
                'condition': weather_context.get('condition', ''),
                'timeOfDay': weather_context.get('timeOfDay', '')
            },
            'sessionId': self.get_current_session_id()
        }

        self.feedback_history.append(feedback_entry)
        self.session_data.append(feedback_entry)

        # Update ML weights based on feedback
        self.update_model_weights(feedback_entry)

        # Update user profile
        self.update_user_profile(feedback_entry)

        return feedback_entry

    def update_model_weights(self, feedback_entry: Dict[str, Any]) -> None:
        """Update ML model weights using reinforcement learning"""
        recommendation = feedback_entry['recommendation']
        feedback = feedback_entry['feedback']
        weather_context = feedback_entry['weatherContext']

        # Calculate reward signal
        reward = self.calculate_reward(feedback)

        # Extract features
        features = {
            'temperature': self.normalize_temperature(weather_context['temperature']),
            'isRainy': 1 if 'rain' in weather_context['condition'].lower() else 0
        }

        # Update weight for this specific recommendation pattern
        feature_key = f"{recommendation['category']}_{int(features['temperature'] * 10)}_{features['isRainy']}"
        current_weight = self.ml_engine.model_weights.get(feature_key, 1)

        # Gradient descent update
        new_weight = current_weight + self.learning_rate * reward
        self.ml_engine.model_weights[feature_key] = max(0, min(2, new_weight))  # Clamp between 0 and 2

        # Also update related patterns (generalization)
        self.update_related_weights(feature_key, reward * 0.5)

    @staticmethod
    def calculate_reward(feedback: str) -> float:
        """Calculate reward signal from user feedback"""
        rewards = {
            'worn': 1.0,  # Strongest positive signal
            'like': 0.5,
            'ignored': -0.2,  # Mild negative signal
            'dislike': -0.5,
            'inappropriate': -1.0  # Strongest negative signal
        }
        return rewards.get(feedback, 0)

    def update_related_weights(self, feature_key: str, adjusted_reward: float) -> None:
        """Update related weights for generalization"""
        parts = feature_key.split('_')
        if len(parts) != 3:
            return

        category, temp_bucket, is_rainy = parts
        temp_num = int(temp_bucket)

        # Update adjacent temperature buckets
        for offset in [-1, 1]:
            related_key = f"{category}_{temp_num + offset}_{is_rainy}"
            current_weight = self.ml_engine.model_weights.get(related_key, 1)
            decay = 0.5 if abs(offset) == 1 else 0.25

            new_weight = current_weight + self.learning_rate * adjusted_reward * decay
            self.ml_engine.model_weights[related_key] = max(0, min(2, new_weight))

    def update_user_profile(self, feedback_entry: Dict[str, Any]) -> None:
        """Update user profile with preference learning"""
        recommendation = feedback_entry['recommendation']
        feedback = feedback_entry['feedback']
        profile = self.ml_engine.user_profile

        # Update category preferences
        if 'categoryPreferences' not in profile:
            profile['categoryPreferences'] = {}

        category_pref = profile['categoryPreferences'].get(recommendation['category'], 1)
        category_reward = self.calculate_reward(feedback) * 0.5
        profile['categoryPreferences'][recommendation['category']] = max(
            0, min(2, category_pref + self.learning_rate * category_reward)
        )

        # Update item preferences
        if 'itemPreferences' not in profile:
            profile['itemPreferences'] = {}

        item_pref = profile['itemPreferences'].get(recommendation['item'], 1)
        item_reward = self.calculate_reward(feedback)
        profile['itemPreferences'][recommendation['item']] = max(
            0, min(2, item_pref + self.learning_rate * item_reward)
        )

        # Store feedback in profile
        if 'feedbackHistory' not in profile:
            profile['feedbackHistory'] = []

        profile['feedbackHistory'].append({
            'timestamp': feedback_entry['timestamp'],
            'item': recommendation['item'],
            'feedback': feedback
        })

        # Keep only last 200 feedback entries
        if len(profile['feedbackHistory']) > 200:
            profile['feedbackHistory'] = profile['feedbackHistory'][-200:]

    def get_personalized_insights(self) -> Dict[str, Any]:
        """Get personalized insights based on learning"""
        if len(self.feedback_history) < 10:
            return {
                'dataPoints': len(self.feedback_history),
                'message': 'Keep providing feedback to get personalized insights!',
                'insights': []
            }

        insights = []
        profile = self.ml_engine.user_profile

        # Analyze category preferences
        category_prefs = profile.get('categoryPreferences', {})
        top_categories = sorted(
            category_prefs.items(),
            key=lambda x: x[1],
            reverse=True
        )[:3]

        if top_categories:
            insights.append({
                'type': 'preference',
                'title': 'Your Favorite Categories',
                'description': ', '.join(
                    f"{cat} ({round(score * 50)}% preference)"
                    for cat, score in top_categories
                )
            })

        # Analyze item preferences
        item_prefs = profile.get('itemPreferences', {})
        loved_items = [item for item, score in item_prefs.items() if score > 1.3]

        if loved_items:
            insights.append({
                'type': 'loved',
                'title': 'Items You Love',
                'description': ', '.join(loved_items[:5])
            })

        disliked_items = [item for item, score in item_prefs.items() if score < 0.7]

        if disliked_items:
            insights.append({
                'type': 'disliked',
                'title': 'Items You Avoid',
                'description': ', '.join(disliked_items[:5])
            })

        # Analyze weather patterns
        weather_patterns = self.analyze_weather_patterns()
        if weather_patterns:
            insights.append(weather_patterns)

        return {
            'dataPoints': len(self.feedback_history),
            'message': f"We've learned from {len(self.feedback_history)} of your feedback entries!",
            'insights': insights,
            'learningProgress': self.calculate_learning_progress()
        }

    def analyze_weather_patterns(self) -> Optional[Dict[str, str]]:
        """Analyze user's weather-based patterns"""
        recent_feedback = self.feedback_history[-50:]

        cold_weather_feedback = [
            f for f in recent_feedback
            if f['weatherContext']['temperature'] < 10
        ]

        warm_weather_feedback = [
            f for f in recent_feedback
            if f['weatherContext']['temperature'] > 20
        ]

        if len(cold_weather_feedback) > 5:
            cold_likes = sum(
                1 for f in cold_weather_feedback
                if f['feedback'] in ['like', 'worn']
            )
            cold_preference = cold_likes / len(cold_weather_feedback)

            if cold_preference > 0.7:
                return {
                    'type': 'weather_pattern',
                    'title': 'Cold Weather Preference',
                    'description': 'You seem to appreciate our cold weather recommendations!'
                }

        if len(warm_weather_feedback) > 5:
            warm_likes = sum(
                1 for f in warm_weather_feedback
                if f['feedback'] in ['like', 'worn']
            )
            warm_preference = warm_likes / len(warm_weather_feedback)

            if warm_preference > 0.7:
                return {
                    'type': 'weather_pattern',
                    'title': 'Warm Weather Preference',
                    'description': 'You love our warm weather recommendations!'
                }

        return None

    def calculate_learning_progress(self) -> Dict[str, Any]:
        """Calculate overall learning progress"""
        total_feedback = len(self.feedback_history)

        if total_feedback < 10:
            return {'level': 'beginner', 'percentage': 10}
        elif total_feedback < 30:
            return {'level': 'learning', 'percentage': 30}
        elif total_feedback < 50:
            return {'level': 'intermediate', 'percentage': 50}
        elif total_feedback < 100:
            return {'level': 'advanced', 'percentage': 70}
        else:
            return {'level': 'expert', 'percentage': 90}

    def get_session_analytics(self) -> Optional[Dict[str, Any]]:
        """Get session-based analytics"""
        total_recommendations = len(self.session_data)

        if total_recommendations == 0:
            return None

        liked = sum(
            1 for f in self.session_data
            if f['feedback'] in ['like', 'worn']
        )

        disliked = sum(
            1 for f in self.session_data
            if f['feedback'] in ['dislike', 'inappropriate']
        )

        return {
            'totalRecommendations': total_recommendations,
            'liked': liked,
            'disliked': disliked,
            'ignored': total_recommendations - liked - disliked,
            'satisfactionRate': liked / total_recommendations
        }

    def reset_session(self) -> None:
        """Reset session data (call when app restarts)"""
        self.session_data = []
        self._session_id = None

    def export_user_data(self) -> Dict[str, Any]:
        """Export user data for analysis or backup"""
        return {
            'feedbackHistory': self.feedback_history,
            'userProfile': self.ml_engine.user_profile,
            'modelWeights': self.ml_engine.model_weights,
            'exportDate': datetime.now().isoformat(),
            'dataPoints': len(self.feedback_history)
        }

    def import_user_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Import user data (for restore or migration)"""
        if 'feedbackHistory' in data:
            self.feedback_history = data['feedbackHistory']

        if 'userProfile' in data:
            self.ml_engine.user_profile = data['userProfile']

        if 'modelWeights' in data:
            self.ml_engine.model_weights = data['modelWeights']

        return {
            'success': True,
            'dataPoints': len(self.feedback_history)
        }

    def clear_all_data(self) -> Dict[str, Any]:
        """Clear all learning data (reset to defaults)"""
        self.feedback_history = []
        self.session_data = []
        self.ml_engine.user_profile = {
            'categoryPreferences': {},
            'itemPreferences': {},
            'feedbackHistory': []
        }
        self.ml_engine.model_weights = {}

        return {'success': True, 'message': 'All learning data cleared'}

    # Helper functions
    def get_current_session_id(self) -> str:
        """Get or create current session ID"""
        if not self._session_id:
            self._session_id = f"session_{int(datetime.now().timestamp() * 1000)}_{uuid.uuid4().hex[:9]}"
        return self._session_id

    @staticmethod
    def normalize_temperature(temp: float) -> float:
        """Normalize temperature"""
        return (temp + 10) / 50
